From 55405d493a0b63aab8fffbe4942fcd3a3ffa20dd Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 16 Jan 2019 17:34:48 +0200
Subject: [PATCH 0898/1051] net: mvpp2: rx dma-sync unmap prefetch adjust
 sequence

This patch adjusts the RX dma-sync/unmap/prefetch with next points:
1). Although we need dma-sync for a Packet-size which is less than
rx-buffer-size it must be done for the SAME size as in map/unmap.
2). The prefetch() instruction is for CPU and therefore should be
done after unmap ~ mapToCPU.

Change-Id: I76074afc06c95fd58b433f9297800492ae02455c
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/2282
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 2b45addc231c..4868ab9c5683 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3545,17 +3545,19 @@ static int mvpp2_rx(struct mvpp2_port *port, struct napi_struct *napi,
 		else
 			frag_size = bm_pool->frag_size;
 
-		/* _sync_ for coherency (_unmap_ is asynchroneous) */
+		/* _sync_ for coherency (_unmap_ is asynchroneous).
+		 * _sync_ should be done for the SAME size as in map/unmap.
+		 * The prefetch is for CPU and should be after unmap ~ mapToCPU
+		 */
 		if (rx_todo == 1)
 			dma_sync_single_for_cpu(dev->dev.parent, dma_addr,
-						MVPP2_RX_BUF_SIZE(rx_bytes),
+						bm_pool->buf_size,
 						DMA_FROM_DEVICE);
-
-		prefetch(data + NET_SKB_PAD); /* packet header */
-
 		dma_unmap_single(dev->dev.parent, dma_addr,
 				 bm_pool->buf_size, DMA_FROM_DEVICE);
 
+		prefetch(data + NET_SKB_PAD); /* packet header */
+
 		skb = mvpp2_build_skb(data, frag_size,
 				      napi, port, rx_status, rxq->id, bm_pool);
 		if (!skb) {
-- 
2.17.1

